src/yunohost/permission.py

```diff
@@ -0,0 +1,513 @@
+
+    raise YunohostError('permission_creation_failed')
+
+
+@is_unit_operation(['permission','app'])
+def permission_update(operation_logger, auth, app, permission, add_url=None, remove_url=None, sync_perm=True):
+    """
+    Update a permission for a specific application
+
+    Keyword argument:
+        app            -- an application OR sftp, xmpp (metronome), mail
+        permission     -- name of the permission ("main" by default)
+        add_url        -- Add a new url for a permission
+        remove_url     -- Remove a url for a permission
+
+    """
+    from yunohost.domain import _normalize_domain_path
+
+    permission_name = str(permission + '.' + app) # str(...) Fix encoding issue
+
+    # Populate permission informations
+    result = auth.search(base='ou=permission,dc=yunohost,dc=org',
+                         filter='cn=' + permission_name, attrs=['URL'])
+    if not result:
+        raise YunohostError('permission_not_found', permission=permission, app=app)
+    permission_obj = result[0]
+
+    if not 'URL' in permission_obj:
+        permission_obj['URL'] = []
+
+    url = set(permission_obj['URL'])
+
+    if add_url:
+        for u in add_url:
+            domain = u[:u.index('/')]
+            path = u[u.index('/'):]
+            domain, path = _normalize_domain_path(domain, path)
+            url.add(domain + path)
+    if remove_url:
+        for u in remove_url:
+            domain = u[:u.index('/')]
+            path = u[u.index('/'):]
+            domain, path = _normalize_domain_path(domain, path)
+            url.discard(domain + path)
+
+    if url == set(permission_obj['URL']):
+        logger.warning(m18n.n('permission_update_nothing_to_do'))
+        return user_permission_list(auth, app, permission)
+
+    operation_logger.start()
+    if auth.update('cn=%s,ou=permission' % permission_name, {'cn':permission_name, 'URL': url}):
+        if sync_perm:
+            permission_sync_to_user(auth)
+        logger.success(m18n.n('permission_updated', permission=permission, app=app))
+        return user_permission_list(auth, app, permission)
+
+    raise YunohostError('premission_update_failed')
+
+
+@is_unit_operation(['permission','app'])
+def permission_remove(operation_logger, auth, app, permission, force=False, sync_perm=True):
+    """
+    Remove a permission for a specific application
+
+    Keyword argument:
+        app        -- an application OR sftp, xmpp (metronome), mail
+        permission -- name of the permission ("main" by default)
+
+    """
+
+    if permission == "main" and not force:
+        raise YunohostError('remove_main_permission_not_allowed')
+
+    operation_logger.start()
+    if not auth.remove('cn=%s,ou=permission' % str(permission + '.' + app)):
+        raise YunohostError('permission_deletion_failed', permission=permission, app=app)
+    if sync_perm:
+        permission_sync_to_user(auth)
+    logger.success(m18n.n('permission_deleted', permission=permission, app=app))
```

Comment:
I apparently triggered a bug when installing then removing an app (well actually two, but should work with one)

After removal of an app, there still are some references in the 'all_users' group (and probably in other group if those existed I guess). Spotted this thanks to the `check_LDAP_db_integrity` test :wink: 

c.f. the references to h5ai and nextcloud in this search : 

```
(Pdb) pp auth.search('ou=groups,dc=yunohost,dc=org', filter='(objectclass=groupOfNamesYnh)', attrs=['cn', 'member', 'memberUid', 'permission'])
[{'cn': ['all_users'],
  'member': ['uid=alice,ou=users,dc=yunohost,dc=org',
             'uid=jack,ou=users,dc=yunohost,dc=org',
             'uid=bob,ou=users,dc=yunohost,dc=org'],
  'memberUid': ['bob', 'alice', 'jack'],
  'permission': ['cn=main.h5ai,ou=permission,dc=yunohost,dc=org',
                 'cn=main.nextcloud,ou=permission,dc=yunohost,dc=org',
                 'cn=main.mail,ou=permission,dc=yunohost,dc=org',
                 'cn=main.metronome,ou=permission,dc=yunohost,dc=org']},
 {'cn': ['alice'], 'member': ['uid=alice,ou=users,dc=yunohost,dc=org']},
 {'cn': ['bob'], 'member': ['uid=bob,ou=users,dc=yunohost,dc=org']},
 {'cn': ['jack'], 'member': ['uid=jack,ou=users,dc=yunohost,dc=org']},
 {'cn': ['dev'],
  'member': ['uid=alice,ou=users,dc=yunohost,dc=org'],
  'memberUid': ['alice']},
 {'cn': ['apps'],
  'member': ['uid=bob,ou=users,dc=yunohost,dc=org'],
  'memberUid': ['bob']}]
```

Not sure how that happens, maybe some stuff not cleared correctly in `permission_sync_to_user`

Reply:
Tried with my_webapp and didn't see this issue.

Are you able to reproduce this bug ?

Reply:
I think I was able to retrigger it using the following steps : 
- Setup a fresh install then pull this PR (and run the migration etc)
- Install Nextcloud
- Run the `backup_and_restore_all_sys` test : `./ynh-dev test yunohost/backuprestore:backup_only_ldap`

Reply:
I've probably found the bug :smile: 

The problem is that actually when we restore only the system (without the app) we only restore the database **without** the permission (because the backup the database without the app permission). The problem is that if we have an app which is installed (before the restoration) and which is not restored, the permission in the "old" database will be replaced by the new database which don't contains any permission !! It's an edge case but when I'm seeing this bug I'm really thinking that the restoration need to be managed more cleanly.

The other problem is that we use in some place the `slapadd` command which is really bad if we wan't to keep the database integrity. In the man of `slapadd` it's written this:
> Your slapd(8) should not be running when you do this to ensure consistency of the database.

So I purpose this solution of that:

- In the backup : save all content of the database
- In the slapd restoration we do this:
  - Save the permission
  - Restore all content of the database (with `slapadd`, but we can because It's all the database so we shouln't have any problem with integrity).
  - Remove all permission in the database
  - Restore the permission which was saved (with `ldapadd` (wich keep the integrity) or with python) for only the apps which is already installed.

With this we should keep a better integrity of the database.

Reply:
So, now it should be fixed by f839ec51

Reply:
Looks like it's fixed indeed, thanks !